src/test/java/com/atlassian/db/replica/internal/state/ConnectionStateTest.java

```diff
@@ -163,4 +165,87 @@ public void shouldChangeStateFromMainToClosed() throws SQLException {
         verify(stateListener).transition(REPLICA, CLOSED);
     }
 
+    @Test
+    public void shouldChangeStateFromNoConnectionToReadFromMainConnection() throws SQLException {
+        final StateListener stateListener = mock(StateListener.class);
+        final ConnectionProviderMock connectionProvider = new ConnectionProviderMock();
+        final Connection connection = DualConnection.builder(
+            connectionProvider,
+            permanentInconsistency().build()
+        ).stateListener(stateListener)
+            .build();
+
+        connection.prepareStatement(SIMPLE_QUERY).executeQuery();
+
+        verify(stateListener).transition(NOT_INITIALISED, READ_FROM_MAIN);
+    }
+
+    @Test
+    public void shouldChangeStateFromReplicaConnectionToReadFromMainConnection() throws SQLException {
+        final StateListener stateListener = mock(StateListener.class);
+        final ConnectionProviderMock connectionProvider = new ConnectionProviderMock();
+        final Connection connection = DualConnection.builder(
+            connectionProvider,
+            new CircularConsistency.Builder(ImmutableList.of(true, false)).build()
+        ).stateListener(stateListener)
+            .build();
+        connection.prepareStatement(SIMPLE_QUERY).executeQuery();
+        Mockito.reset(stateListener);
+
+        connection.prepareStatement(SIMPLE_QUERY).executeQuery();
+
+        verify(stateListener).transition(REPLICA, READ_FROM_MAIN);
+    }
+
+    @Test
+    public void shouldChangeStateFromReadFromMainConnectionToReplicaConnection() throws SQLException {
```

Comment:
This redescribes the assertion. But why or when should it change the state in that way?

Reply:
`shouldChangeStateFromReadFromMainConnectionToReplicaConnectionWhenNoWritesOnMain` will be better?

Reply:
Nice! Now I know the what **and** the why.
Lemme try some rewording iterations:
* firstly, drop redundancy: `shouldSwitchFromMainToReplicaWhenNoWritesOnMain`
* secondly some semantic compression: `shouldForgiveReplicaWhenNoWritesOnMain`
* thirdly, reducing negation:  `shouldForgiveReplicaIfItCatchesUpOnReads`.

And counterpart test case (if it exists): `shouldAvoidReplicaDueToPossibleTransactions`.